agent/jvm/src/main/java/org/jolokia/jvmagent/security/KeyStoreUtil.java

```diff
@@ -137,24 +147,48 @@ private static Object createKeyPair() throws NoSuchAlgorithmException, NoSuchPro
         return keypair;
     }
 
-    private static X509Certificate getSelfCertificate(Object keypair, X500Name x500Name, Date date, long l) {
-        return (X509Certificate) ClassUtil.applyMethod(keypair, "getSelfCertificate", x500Name, date, l);
+    private static X509Certificate getSelfCertificate(Object keypair, Object x500Name, Date date, long l) {
+        final Class<?> clazz;
+        if (ClassUtil.checkForClass(X500_NAME_SUN)) {
+            clazz = ClassUtil.classForName(X500_NAME_SUN, false);
+        } else {
+            clazz = ClassUtil.classForName(X500_NAME_IBM, false);
+        }
+
+        try {
+            final Method selfCertMethod = keypair.getClass().getDeclaredMethod("getSelfCertificate", clazz, Date.class, long.class);
```

Comment:
I wonder, whether the original `ClassUtil.applyMethod()` wouldn't work here as well ? The incoming `x500Name` object already would be already of the proper IBM / Oracle type and `applyMethod()` should work in both cases. Have you tried it and got some error ?

Reply:
Yes, it fails because the applyMethod would search for a method-signature that does not exist (if I remember correctly right now it'd say that there is no method that takes Object, Date, long). Of course it would be possible to modify applyMethod but at that moment I thought it's less invasive in your code to simply do it this way.